File: src/main/java/discretemath/structure/sequence/Sequence.java
package discretemath.structure.sequence;

import discretemath.structure.exception.NotSummableException;
import discretemath.structure.exception.TermNotDefinedException;
import discretemath.structure.function.Function;

public interface Sequence<T> extends Function<Integer, T> {

	T term(int n) throws TermNotDefinedException;

	T sum(int from, int to) throws NotSummableException;
}

File: src/main/scala/discretemath/structure/sequence/Sequence.scala
package discretemath.structure.sequence

import discretemath.common.Adder
import discretemath.structure.exception.NotSummableException
import discretemath.structure.function._

import scala.math.BigInt

trait Sequence[T] extends Function[BigInt, T] {

  @throws[NotSummableException]
  def sum(implicit adder: Adder[T], from: Int, to: Int): T
}